plugin/pkg/proxy/connect.go

```diff
@@ -117,6 +119,31 @@ func (p *Proxy) Connect(ctx context.Context, state request.Request, opts Options
 	for {
 		ret, err = pc.c.ReadMsg()
 		if err != nil {
+			if proto == "udp" {
+				// For UDP, if the error is an overflow, we probably have an upstream misbehaving in some way.
+				// (e.g. sending >512 byte responses without an eDNS0 OPT RR).
+				// Instead of returning an error, return an empty response with TC bit set. This will make the
+				// client retry over TCP (if that's supported) or at least receive a clean
+				// error. The connection is still good so we break before the close.
+				shouldTruncate := false
+
+				// This is to handle a scenario in which upstream sets the TC bit, but doesn't truncate the response
+				// and we get ErrBuf instead of overflow.
+				if _, isDNSErr := err.(*dns.Error); isDNSErr && errors.Is(err, dns.ErrBuf) {
+					shouldTruncate = true
+				} else if strings.Contains(err.Error(), "overflow") {
+					shouldTruncate = true
+				}
```

Comment:
Please make this into a function, and add unit tests for it.

Reply:
Implemented both the suggested changes.